stdlib/public/core/Zip.swift

```diff
@@ -153,5 +151,261 @@ extension Zip2Sequence: Sequence {
   }
 }
 
-// @available(*, deprecated, renamed: "Zip2Sequence.Iterator")
+/// Creates a collection of pairs built out of two underlying collections.
+///
+/// In the `Zip2Collection` instance returned by this function, the elements of
+/// the *i*th pair are the *i*th elements of each underlying collection. The
+/// following example uses the `zip(_:_:)` function to iterate over an array
+/// of strings and a countable range at the same time:
+///
+///     let words = ["one", "two", "three", "four"]
+///     let numbers = 1...4
+///
+///     for (word, number) in zip(words, numbers) {
+///         print("\(word): \(number)")
+///     }
+///     // Prints "one: 1"
+///     // Prints "two: 2
+///     // Prints "three: 3"
+///     // Prints "four: 4"
+///
+/// If the two sequences passed to `zip(_:_:)` are different lengths, the
+/// resulting sequence is the same length as the shorter sequence. In this
+/// example, the resulting array is the same length as `words`:
+///
+///     let naturalNumbers = 1...Int.max
+///     let zipped = Array(zip(words, naturalNumbers))
+///     // zipped == [("one", 1), ("two", 2), ("three", 3), ("four", 4)]
+///
+/// - Parameters:
+///   - collection1: The first collection to zip.
+///   - collection2: The second collection to zip.
+/// - Returns: A sequence of tuple pairs, where the elements of each pair are
+///   corresponding elements of `collection1` and `collection2`.
+@_inlineable // FIXME(sil-serialize-all)
+public func zip<Collection1, Collection2>(
+  _ collection1: Collection1, _ collection2: Collection2
+) -> Zip2Collection<Collection1, Collection2> {
+  return Zip2Collection(collection1, collection2)
+}
+
+public struct Zip2Collection<Collection1 : Collection, Collection2 : Collection> {
+  @_versioned // FIXME(sil-serialize-all)
+  internal let _collection1: Collection1
+  @_versioned // FIXME(sil-serialize-all)
+  internal let _collection2: Collection2
+  
+  /// Creates an instance that makes pairs of elements from `collection1` and
+  /// `collection2`.
+  // FIXME: this should be inlineable, but this results in a compilation error:
```

Comment:
When a non-delegating initializer is inlinable, it exposes the stored property layout of the type. So the type has to be @_fixed_layout, which this one is not.

Reply:
Ahhhh thanks that's what I was missing.

I raised [SR-6759](https://bugs.swift.org/browse/SR-6759) for a clearer diagnostic.